### PR TITLE
Add support for `event::Touch` on web targets

### LINKED ISSUE #1673: Implement touch handling on web target(s)
Adding touch event handling shouldn't be too hard, but it is tightly coupled with mouse handling due to the web using a unified Pointer Event model. Here are my thoughts:

- Grabbing touch input from `PointerEvent` is simple enough. However, I would be concerned about correctly mapping the [`pointerType` ](https://developer.mozilla.org/en-US/docs/Web/API/PointerEvent/pointerType) :
    - `"mouse"` - obviously should be mapped to `CursorMove`, `CursorEntered`, `CursorLeft` and `MouseInput`. In addition, I believe that only pointers with `isPrimary === true` should be reported.
    - `"touch"` - obviously should be mapped to `Touch`.
    - `"pen"` - this is tricky, as `winit` does not have an event for pen/stylus. I suggest that for now we just map pen input the same way as mouse input until `winit` finally supports pen events properly.
    - Empty strings or unknown values - the same as mouse.
- Supporting browsers without `PointerEvent` - I suggest we do **not** add support for `TouchEvent`, because:
    - `PointerEvent` is getting [widespread support](https://caniuse.com/#feat=pointer).
    - `Adding TouchEvent` handling will likely make the code quite complicated.
    - ~~If the user really care about supporting older browsers, they can use a [JavaScript polyfill](https://github.com/jquery/PEP).~~ Turns out the polyfill does not provide `offsetX`/`offsetY` which `winit` is currently using...
- Should touch input also synthesize mouse input?
    - My idea would be that, since some users may not want to go through the trouble of implementing explicit touch handling there should be an option to allow user code to disable touch support and automatically treat the primary touch input as mouse input. Otherwise, touch input will only generate touch events and not mouse events.
    - However, it seems that this is not an option for other platforms. Perhaps it makes more sense to just provide one behaviour, which is for touch input to only generate touch events and not mouse events.

### PR DESCRIPTION
**Pull request by dsluijk**
- [ ] Tested on all platforms changed
- [x] Compilation warnings were addressed
- [x] `cargo fmt` has been run on this branch
- [x] `cargo doc` builds successfully
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

This pull request adds touch support to web targets by the Pointer events API.

This PR also adds a new datatype to the Touch struct, `PointerType`. This indicates which pointer is being used to generate the touch event. As this has been added to the platform-independent Touch struct, all platforms supporting touch had to be changed slightly. I manually checked all implementations and added a None value to all (non-web) Touch structs, but it is possible I missed one.

Closes #1673 .

### CODE REVIEW

File: CHANGELOG.md
Comment (maroider) [inline CHANGELOG.md:4]:
```suggestion
- On Web, added support for `WindowEvent::Touch`.
- Added `Window::focus_window`to bring the window to the front and set input focus.
```

File: src/event.rs
Comment (maroider) [inline src/event.rs:671]:
```suggestion
```

Is it truly necessary to expose this now?
Adding a richer pointer API has been discussed in #99 and #336, but I don't think this is the right time (or place) to add it.

**Review by maroider — CHANGES_REQUESTED** [review]
You indicated that you want to help out with WASM, @FuriouZz, and I think helping review and test this (and other) web PRs would be a good start. You don't *need* to be a maintainer to review and/or test PRs.

File: CHANGELOG.md
Comment (maroider) [reply CHANGELOG.md]: You still need to make it say `WindowEvent::Touch` rather than `Event::Touch`

Comment (dsluijk) [reply CHANGELOG.md]: Sorry, read over that change. Fixed it now!

File: src/event.rs
Comment (dsluijk) [reply src/event.rs]: Fair enough, I've remove it and any reference to it for now.

File: src/platform_impl/web/event_loop/window_target.rs
Comment (FuriouZz) [inline src/platform_impl/web/event_loop/window_target.rs:242]: It is currently possible to implement the force on Web for Apple Devices ([doc](https://developer.apple.com/library/archive/documentation/AppleApplications/Conceptual/SafariJSProgTopics/RespondingtoForceTouchEventsfromJavaScript.html)). I think it can be nice to have a complete implementation.

**Review by FuriouZz — CHANGES_REQUESTED** [review]
The implementation sounds good to me. The force touch is currently implemented on iOS and it is possible on the web (see my comment). I prefer to validate this PR with its implementation.
Just asking @maroider, this may be the subject of another Issue/PR : winit does have any plans to support multi-touch ?

**Comment by maroider** [comment]
> Just asking @maroider, this may be the subject of another Issue/PR : winit does have any plans to support multi-touch ?

I believe an API for that (and a pointer event API) was explored in #1374, and potentially in  #99 and/or #336,

File: src/platform_impl/web/stdweb/canvas.rs
Comment (maroider) [inline src/platform_impl/web/stdweb/canvas.rs:326]: You should remove this. Support for `stdweb` is being removed in the very near future (#1941).

File: src/platform_impl/web/web_sys/canvas/pointer_handler.rs
Comment (maroider) [inline src/platform_impl/web/web_sys/canvas/pointer_handler.rs:149]: I'm not sure if panicking is the correct response to failure here. I'd honestly be surprised myself if my (hypothetical) application crashed because the browser didn't let Winit use `Element.setPointerCapture` for some reason.

Also: I don't quite understand why using `Element.setPointerCapture` is necessary here in the first place.

**Review by maroider — CHANGES_REQUESTED** [review]
Is there anything being done in this PR to prevent non-touch pointer events being reported as `WindowEvent::Touch`?

File: src/platform_impl/web/web_sys/canvas/pointer_handler.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas/pointer_handler.rs:33]: Rename to `on_touch_(move|down|up|cancel)`

File: src/platform_impl/web/web_sys/canvas/pointer_handler.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas/pointer_handler.rs:122]: The handler must be called only when `event.pointer_type()` is equal to `PointerType::Touch`

**Review by FuriouZz — CHANGES_REQUESTED** [review]

File: src/platform_impl/web/stdweb/canvas.rs
Comment (dsluijk) [reply src/platform_impl/web/stdweb/canvas.rs:326]: I've added this as the CI was failing, but if it's good to merge with stdweb failing then that's even better. I'll revert this file

File: src/platform_impl/web/web_sys/canvas/pointer_handler.rs
Comment (dsluijk) [reply src/platform_impl/web/web_sys/canvas/pointer_handler.rs:149]: I've copied this from the `on_mouse_press` event in the same file. These events are very similar, so I thought this was required. I'll remove it if it's not necessary.

File: src/platform_impl/web/web_sys/canvas/pointer_handler.rs
Comment (dsluijk) [reply src/platform_impl/web/web_sys/canvas/pointer_handler.rs:122]: Changed!

File: src/platform_impl/web/web_sys/canvas/pointer_handler.rs
Comment (dsluijk) [reply src/platform_impl/web/web_sys/canvas/pointer_handler.rs:33]: Renamed

File: src/platform_impl/web/event_loop/window_target.rs
Comment (dsluijk) [reply src/platform_impl/web/event_loop/window_target.rs:242]: I think this is currently out of scope for this PR, but feel free to add it in another one!

**Comment by dsluijk** [comment]
> Is there anything being done in this PR to prevent non-touch pointer events being reported as `WindowEvent::Touch`?

I've added a guard for it now.

**Comment by FuriouZz** [comment]
@maroider Just noticied that PR #1900 already exist for event::Touch support created by @zicklag. That PR use touch event and support multitouch.

**Comment by maroider** [comment]
Well, I'll be damned. I had completely forgotten about that PR.

**Comment by dsluijk** [comment]
Do note that that PR uses touch events instead of pointers, as was preferred in #1673

**Comment by maroider** [comment]
> Do note that that PR uses touch events instead of pointers, as was preferred in #1673 

We'll go ahead with this PR in favour of #1900, in that case.

**Comment by zicklag** [comment]
So I should close #1900 to avoid confusion, right?

**Comment by maroider** [comment]
Sure.

#1941 got merged, @dsluijk, so you'll want to rebase onto `master`.

**Comment by dsluijk** [comment]
> #1941 got merged, @dsluijk, so you'll want to rebase onto `master`.

Done! I think it's ready to merge now, is there anything what still needs to be updated?

**Review by maroider — APPROVED** [review]
I'd like to do a real-world test, but this should otherwise be ready to be merged, despite how CI is a bit wonky ATM.

I'd also like a final sign-off from @FuriouZz.

**Comment by zicklag** [comment]
@maroider I can try this out, if you want before you merge it. I'm currently having to use my fork in order to power my [Bounty Bros.](https://katharostech.com/post/bounty-bros-on-web) game touch support on web so I have a perfect test case. I'll try to test it within the next ~12 hours ( probably less ) and comment here with the result.

**Comment by maroider** [comment]
Sounds like a plan, @zicklag.

File: src/platform_impl/web/event_loop/window_target.rs
Comment (FuriouZz) [inline src/platform_impl/web/event_loop/window_target.rs:235]:
```suggestion
        canvas.on_touch_move(move |device_id, location| {
```

File: src/platform_impl/web/event_loop/window_target.rs
Comment (FuriouZz) [inline src/platform_impl/web/event_loop/window_target.rs:249]:
```suggestion
        canvas.on_touch_down(move |device_id, location| {
```

File: src/platform_impl/web/event_loop/window_target.rs
Comment (FuriouZz) [inline src/platform_impl/web/event_loop/window_target.rs:263]:
```suggestion
        canvas.on_touch_up(move |device_id, location| {
```

File: src/platform_impl/web/event_loop/window_target.rs
Comment (FuriouZz) [inline src/platform_impl/web/event_loop/window_target.rs:277]:
```suggestion
        canvas.on_touch_cancel(move |device_id, location| {
```

File: src/platform_impl/web/web_sys/canvas.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas.rs:250]:
```suggestion
    pub fn on_touch_move<F>(&mut self, handler: F)
```

File: src/platform_impl/web/web_sys/canvas.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas.rs:255]:
```suggestion
            MouseState::HasPointerEvent(h) => h.on_touch_move(&self.common, handler),
```

File: src/platform_impl/web/web_sys/canvas.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas.rs:260]:
```suggestion
    pub fn on_touch_down<F>(&mut self, handler: F)
```

File: src/platform_impl/web/web_sys/canvas.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas.rs:265]:
```suggestion
            MouseState::HasPointerEvent(h) => h.on_touch_down(&self.common, handler),
```

File: src/platform_impl/web/web_sys/canvas.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas.rs:270]:
```suggestion
    pub fn on_touch_up<F>(&mut self, handler: F)
```

File: src/platform_impl/web/web_sys/canvas.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas.rs:275]:
```suggestion
            MouseState::HasPointerEvent(h) => h.on_touch_up(&self.common, handler),
```

File: src/platform_impl/web/web_sys/canvas.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas.rs:280]:
```suggestion
    pub fn on_touch_cancel<F>(&mut self, handler: F)
```

File: src/platform_impl/web/web_sys/canvas.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas.rs:285]:
```suggestion
            MouseState::HasPointerEvent(h) => h.on_touch_cancel(&self.common, handler),
```

File: src/platform_impl/web/web_sys/canvas/pointer_handler.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas/pointer_handler.rs:114]:
```suggestion
    pub fn on_touch_move<F>(&mut self, canvas_common: &super::Common, mut handler: F)
```

File: src/platform_impl/web/web_sys/canvas/pointer_handler.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas/pointer_handler.rs:136]:
```suggestion
    pub fn on_touch_down<F>(&mut self, canvas_common: &super::Common, mut handler: F)
```

File: src/platform_impl/web/web_sys/canvas/pointer_handler.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas/pointer_handler.rs:158]:
```suggestion
    pub fn on_touch_up<F>(&mut self, canvas_common: &super::Common, mut handler: F)
```

File: src/platform_impl/web/web_sys/canvas/pointer_handler.rs
Comment (FuriouZz) [inline src/platform_impl/web/web_sys/canvas/pointer_handler.rs:180]:
```suggestion
    pub fn on_touch_cancel<F>(&mut self, canvas_common: &super::Common, mut handler: F)
```

**Review by FuriouZz — CHANGES_REQUESTED** [review]
Just some reword from `pointer` to `touch`

**Comment by zicklag** [comment]
I'll be investigating, more, but for some reason when testing with my game it doesn't seem to be picking up tap-and-drags, when inputing through the Chrome browser's mobile viewer. I think maybe it's not holding down the touch or something. I'll debug more in a second to see what the difference in the the sent events are between my fork and this PR.

**Comment by Friz64** [comment]
> I'll be investigating, more, but for some reason when testing with my game it doesn't seem to be picking up tap-and-drags, when inputing through the Chrome browser's mobile viewer.

I am experiencing the same thing. I was using zicklag's PR (#1900) earlier, which reports touch events as `CursorMoved`. I prefer this as it allows me to use only one event on all platforms as was also mentioned in #1673.

**Comment by maroider** [comment]
> I'll be investigating, more, but for some reason when testing with my game it doesn't seem to be picking up tap-and-drags, when inputing through the Chrome browser's mobile viewer.

This seems to be an oversight on my end. I thought the call to `set_pointer_capture` [wasn't removed](https://github.com/rust-windowing/winit/pull/1945#discussion_r636340068), and I think removing that call is what's causing this issue.

**Comment by zicklag** [comment]
Here's the event log I'm getting when touch-and-dragging if this helps:

image

It seems to capture the first so much of a move before getting a touch-cancelled event.

**Comment by maroider** [comment]
Did you try it with `set_pointer_capture` re-added in the correct place, @zicklag?

**Comment by zicklag** [comment]
That didn't seem to fix it, but in my latest tests it doesn't seem like the event handler is getting run at all, maybe. I'm looking around.

**Comment by FuriouZz** [comment]
The [problem](https://stackoverflow.com/questions/48124372/pointermove-event-not-working-with-touch-why-not) seems to come from the browser canceling pointermove event. Setting `html, body { touch-action: none; }` fix the issue. Does exist an alternative of that fix?
@zicklag @maroider

**Comment by maroider** [comment]
Does it work if you only apply the style to the canvas? If so, we can apply the style programmatically.

**Comment by FuriouZz** [comment]
It works on my sample 
> canvas.style().set_property("touch-action", "none");

> Does it work if you only apply the style to the canvas? If so, we can apply the style programmatically.
> We'll probably also want to look into what using `preventDefault()` looks like, although I don't think we can get away with using it unconditionally.

I tried that, but it does not work.

**Comment by zicklag** [comment]
> canvas.style().set_property("touch-action", "none");

That worked for me, too. Looks like that might be all we need.

**Comment by FuriouZz** [comment]
> canvas.style().set_property("touch-action", "none");

Must be added [here](https://github.com/rust-windowing/winit/blob/master/src/platform_impl/web/web_sys/canvas.rs#L68) and maybe [here](https://github.com/rust-windowing/winit/blob/master/src/platform/web.rs#L26) too.

**Comment by ArturKovacs** [comment]
If I'm not mistaken then the previous message was directed to you, @dsluijk. If you don't have the opportunity to continue working on this, feel free to indicate that.

**Comment by SamiPirbay** [comment]
It works but the origin (0, 0) is at the top-left corner, with Y going downwards.

**Comment by FuriouZz** [comment]
Up please

**Comment by johanhelsing** [comment]
Is there something I can do to help this one along?

I backported the PR to winit 0.24 (that was what bevy 0.5 was on) and tested it with [one of my games](https://helsing.studio/tails), and it worked great. Would be nice to have it merged.

@dsluijk, if you don't have the time, I can try to see if I can resolve the conflicts and flip the y-axis.

**Comment by madsmtm** [comment]
https://github.com/rust-windowing/winit/pull/2188 has been merged, which fixes this as well.

**Comment by FuriouZz** [comment]
Finally!

**Pull request closed without merging**
